tests/rptest/tests/delete_records_test.py

```diff
@@ -775,3 +783,153 @@ class PartitionMoveExceptionReporter:
+        topic = TopicSpec(name=TEST_TOPIC_NAME,
+                          partition_count=1,
+                          replication_factor=3)
+        self.client.create_topic(topic)
+
+        producer = KgoVerifierProducer(self.ctx,
+                                       self.redpanda,
+                                       TEST_TOPIC_NAME,
+                                       msg_size=1,
+                                       msg_count=10)
+
+        producer.start()
+        producer.wait()
+
+        # Make an initial DeleteRecords request to bump up the start offset.
+        delete_records_tpos = [(TEST_TOPIC_NAME, 0, 5)]
+        delete_records_json = self.make_delete_records_for_cli(
+            delete_records_tpos)
+        delete_records_result = self.execute_delete_records(
+            delete_records_json)[0]
+
+        # Assert no error, should see the new_start_offset == 5.
+        assert delete_records_result.error_msg == ""
+        new_start_offset = delete_records_result.new_start_offset
+        assert new_start_offset == 5
+
+        def do_test_body(truncate_offset, expect_error,
+                         expected_new_start_offset):
+            delete_records_tpos = [(TEST_TOPIC_NAME, 0, truncate_offset)]
+            delete_records_json = self.make_delete_records_for_cli(
+                delete_records_tpos)
+            delete_records_result = self.execute_delete_records(
+                delete_records_json)[0]
+
+            if expect_error:
+                exception_name = "org.apache.kafka.common.errors.OffsetOutOfRangeException"
+                assert exception_name in delete_records_result.error_msg
+            else:
+                assert delete_records_result.error_msg == ""
+                assert delete_records_result.new_start_offset == expected_new_start_offset
+
+            if expect_error:
+                with expect_exception(
+                        RpkException,
+                        lambda e: "OFFSET_OUT_OF_RANGE" in str(e)):
+                    self.rpk.trim_prefix(TEST_TOPIC_NAME, truncate_offset, [0])
+            else:
+                trim_prefix_result = self.rpk.trim_prefix(
+                    TEST_TOPIC_NAME, truncate_offset, [0])[0]
+                assert trim_prefix_result.error_msg == ""
+                assert trim_prefix_result.new_start_offset == expected_new_start_offset
+
+        truncate_offset = None
+        expect_error = False
+        expected_new_start_offset = None
+        if truncate_point == "start_offset":
+            truncate_offset = new_start_offset
+            expected_new_start_offset = truncate_offset
+        elif truncate_point == "below_start_offset":
+            truncate_offset = new_start_offset - 1
+            expect_error = True
+        else:
+            assert False, "unknown truncate point encountered"
+
+        do_test_body(truncate_offset, expect_error, expected_new_start_offset)
```

Comment:
confused as to why this is extracted as an inner function?

Reply:
Cleaned up test by removing inner function.